File: benchmarks/mat4x4.go
package main

import (
	"fmt"
	"math"
	"time"
)

const RUNS = 3
const ITERS = 1000000

type Mat4 [4][4]float64

func mat4Mul(a, b Mat4) Mat4 {
	var r Mat4
	for i := 0; i < 4; i++ {
		for j := 0; j < 4; j++ {
			for k := 0; k < 4; k++ {
				r[i][j] += a[i][k] * b[k][j]
			}
		}
	}
	return r
}

func main() {
	fmt.Println("=== Go Mat4x4 Benchmarks ===")

	b := Mat4{
		{0.8660254037844387, -0.5, 0.0, 0.0},
		{0.5, 0.8660254037844387, 0.0, 0.0},
		{0.0, 0.0, 1.0, 0.0},
		{0.0, 0.0, 0.0, 1.0},
	}

	// Warmup
	a := Mat4{{1, 0, 0, 0}, {0, 1, 0, 0}, {0, 0, 1, 0}, {0, 0, 0, 1}}
	for i := 0; i < ITERS; i++ {
		a = mat4Mul(a, b)
	}
	_ = a

	// Best of 3
	best := int64(math.MaxInt64)
	checksum := 0.0
	for run := 0; run < RUNS; run++ {
		a = Mat4{{1, 0, 0, 0}, {0, 1, 0, 0}, {0, 0, 1, 0}, {0, 0, 0, 1}}
		start := time.Now()
		for i := 0; i < ITERS; i++ {
			a = mat4Mul(a, b)
		}
		elapsed := time.Since(start).Nanoseconds()
		if elapsed < best {
			best = elapsed
			checksum = 0
			for i := 0; i < 4; i++ {
				for j := 0; j < 4; j++ {
					checksum += a[i][j]
				}
			}
		}
	}
	fmt.Printf("Mat4x4 multiply (%d iterations):\n", ITERS)
	fmt.Printf("  Time: %d ms\n", best/1000000)
	fmt.Printf("  Result: %.10f\n", checksum)
}

File: benchmarks/mat4x4.rs
use std::hint::black_box;
use std::time::Instant;

const RUNS: usize = 3;
const ITERS: usize = 1000000;

type Mat4 = [[f64; 4]; 4];

fn mat4_mul(a: &Mat4, b: &Mat4) -> Mat4 {
    let mut r = [[0.0f64; 4]; 4];
    for i in 0..4 {
        for j in 0..4 {
            for k in 0..4 {
                r[i][j] += a[i][k] * b[k][j];
            }
        }
    }
    r
}

fn main() {
    println!("=== Rust Mat4x4 Benchmarks ===");

    let b: Mat4 = [
        [0.8660254037844387, -0.5, 0.0, 0.0],
        [0.5, 0.8660254037844387, 0.0, 0.0],
        [0.0, 0.0, 1.0, 0.0],
        [0.0, 0.0, 0.0, 1.0],
    ];

    // Warmup
    let mut a: Mat4 = [[0.0; 4]; 4];
    a[0][0] = 1.0; a[1][1] = 1.0; a[2][2] = 1.0; a[3][3] = 1.0;
    for _ in 0..ITERS {
        a = mat4_mul(&a, black_box(&b));
    }
    let _ = black_box(a);

    // Best of 3
    let mut best = u128::MAX;
    let mut checksum: f64 = 0.0;
    for _ in 0..RUNS {
        a = [[0.0; 4]; 4];
        a[0][0] = 1.0; a[1][1] = 1.0; a[2][2] = 1.0; a[3][3] = 1.0;
        let start = Instant::now();
        for _ in 0..ITERS {
            a = mat4_mul(&a, black_box(&b));
        }
        let elapsed = start.elapsed().as_nanos();
        if elapsed < best {
            best = elapsed;
            checksum = 0.0;
            for i in 0..4 {
                for j in 0..4 {
                    checksum += a[i][j];
                }
            }
        }
    }
    println!("Mat4x4 multiply ({} iterations):", ITERS);
    println!("  Time: {} ms", best / 1000000);
    println!("  Result: {:.10}", checksum);
}
